rs/bitcoin/ckbtc/minter/src/updates/retrieve_btc.rs

```diff
@@ -299,6 +299,31 @@ pub async fn retrieve_btc_with_approval(
         ));
     }
 
+    let new_kyt_principal = read_state(|s| {
+        s.new_kyt_principal
+            .expect("BUG: upgrade procedure must ensure that the new KYT principal is set")
+            .get()
+            .into()
+    });
+
+    match new_kyt_check_address(new_kyt_principal, args.address.clone()).await {
+        Err(error) => {
+            return Err(RetrieveBtcWithApprovalError::GenericError {
+                error_message: format!("Failed to call KYT canister with error: {:?}", error),
+                error_code: ErrorCode::KytCallFailed as u64,
+            })
+        }
+        Ok(status) => match status {
+            BtcAddressCheckStatus::Tainted => {
```

Comment:
We don't need to charge a KYT fee because the ckBTC minter fee is (more than) high enough if a Bitcoin transaction is made.
However, a cycle drainage attack is possible by making calls that try to retrieve to a tainted address. Shouldn't we deduct a small fee in ckBTC to pay for the cycles cost?

Reply:
The check result is cached. So I guess similar attacks would apply to all update methods that do not charge a fee. Shall we worry about them all?

Reply:
> The check result is cached.

I'm not sure what you mean here. The address is checked by a _query_ call on the KYT canister that checks it against a hard-coded blocklist. Meaning that there is still the cost of the inter-canister query call (from minter to KYT), but is that something we should worry about?

If it's a problem, we could also cache tainted addresses on the minter (to avoid inter-canister calls) if that's something we are really worried about. WDYT?

Reply:
> I'm not sure what you mean here. The address is checked by a query call on the KYT canister that checks it against a hard-coded blocklist. Meaning that there is still the cost of the inter-canister query call (from minter to KYT)

Yes, you are right. I got confused with the deposit case where the utxo check result is being cached locally in the minter.

> If it's a problem, we could also cache tainted addresses on the minter (to avoid inter-canister calls) if that's something we are really worried about. WDYT?

Yes, I agree. I don't think we need to worry about it right now.

Reply:
I didn't know that we are using composite queries already to interact with the KYT canister. In that case, I guess it's fine the way it is - until charging for query calls is introduced.   :slightly_smiling_face: